zerver/lib/actions.py

```diff
@@ -3658,6 +3665,28 @@ def do_delete_message(user_profile: UserProfile, message: Message) -> None:
     move_message_to_archive(message.id)
     send_event(event, ums)
 
+def get_all_streams_message_count() -> int:
+    stat = COUNT_STATS['messages_in_stream:is_bot:day']
+    d = timezone_now() - datetime.timedelta(days=28)
+    return StreamCount.objects.filter(property=stat.property, end_time__gt=d)
+
+def get_stream_message_count(stream_id: int, stream_date_created: datetime.datetime,
+                             message_counts: QuerySet) -> int:
+    count = 0
+    for message_count in message_counts:
+        if message_count.stream_id == stream_id:
+            count += message_count.value
```

Comment:
This loop could get quite expensive on larger realms, since we end up iterating through a (NDays * NStreams) data set once for each stream, for quadratic perf in the number of streams.

We need an interface here where:
(1) The bulk operation, for e.g. `gather_subscriptions_helper`, where we want it for all streams, just processes the full data set once, and returns e.g. a dict mapping {stream_id -> that stream's data, maybe a tuple (days, count) or even just the answer}.  
(2) The single-stream or short-list-of-streams operation, which we need for the `bulk_subscribe` process, works similarly, but only needs to process those `stream_id` values that are actually involved.